amy/emails/actions/instructor_training_completed_not_badged.py

```diff
@@ -0,0 +1,286 @@
+from datetime import date, datetime
+import logging
+
+from django.contrib.contenttypes.models import ContentType
+from django.http import HttpRequest
+from typing_extensions import Unpack
+
+from emails.actions.base_action import BaseAction, BaseActionCancel, BaseActionUpdate
+from emails.models import ScheduledEmail
+from emails.signals import (
+    INSTRUCTOR_TRAINING_COMPLETED_NOT_BADGED_SIGNAL_NAME,
+    Signal,
+    instructor_training_completed_not_badged_remove_signal,
+    instructor_training_completed_not_badged_signal,
+    instructor_training_completed_not_badged_update_signal,
+)
+from emails.types import (
+    InstructorTrainingCompletedNotBadgedContext,
+    InstructorTrainingCompletedNotBadgedKwargs,
+    StrategyEnum,
+)
+from emails.utils import two_months_after
+from workshops.models import Person, TrainingProgress
+
+from .instructor_training_approaching import EmailStrategyException
+
+logger = logging.getLogger("amy")
+
+
+class TrainingCompletionDateException(Exception):
+    pass
+
+
+def find_training_completion_date(person: Person) -> date:
+    """Given a person, find their passed training and related event.
+    Return event end date."""
+    training = TrainingProgress.objects.get(
+        trainee=person, state="p", requirement__name="Training"
+    )
+    event = training.event
+    if not event:
+        raise TrainingCompletionDateException(
+            "Training progress doesn't have an event."
+        )
+
+    training_completed_date = event.end
+    if not training_completed_date:
+        raise TrainingCompletionDateException(
+            "Training progress event doesn't have an end date."
+        )
+
+    return training_completed_date
+
+
+def instructor_training_completed_not_badged_strategy(person: Person) -> StrategyEnum:
+    logger.info(f"Running InstructorTrainingCompletedNotBadged strategy for {person}")
+
+    person_annotated = (
+        Person.objects.annotate_with_instructor_eligibility().get(  # type: ignore
+            pk=person.pk
+        )
+    )
+
+    all_requirements_passed = (
+        bool(person_annotated.passed_training)
+        and bool(person_annotated.passed_get_involved)
+        and bool(person_annotated.passed_welcome)
+        and bool(person_annotated.passed_demo)
+    )
+
+    ct = ContentType.objects.get_for_model(person)  # type: ignore
+    has_email_scheduled = ScheduledEmail.objects.filter(
+        generic_relation_content_type=ct,
+        generic_relation_pk=person.pk,
+        template__signal=INSTRUCTOR_TRAINING_COMPLETED_NOT_BADGED_SIGNAL_NAME,
+        state="scheduled",
+    ).exists()
```

Comment:
We may be adding progress after this email has been sent, since the email is intended to act as a reminder. Could you also check for running/succeeded state and prevent a new email from being created in that case?

Reply:
I see your point.

Potentially, in the worst case scenario, as soon as first email reminder was sent, there's a new training progress passed causing another email to schedule. It would run immediately if the event's end date hadn't change and it's >= 2 months in the past.

But this can also be aligned with what business (@maneesha) wants. I mean, another one or more reminders may not cause harm.

@maneesha Consider the scenario above and let me know if only one reminder should be sent at most.

Reply:
Right now, we only send this email once, at 60 days after training, letting them know their 90 day deadline is approaching.  In our current workflows, no more emails are sent.

This may also be one that is better suited to run via a cronjob rather than the triggers and conditions other emails use.

Reply:
Okay, I'll change the implementation to run only once, and then we'll see after tests how good this works - and optionally change it to cronjob type.